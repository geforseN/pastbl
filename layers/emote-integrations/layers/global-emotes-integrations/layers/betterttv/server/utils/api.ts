import type { BetterTTVApi } from "#integrations_betterttv/api-types";

// LINK: https://betterttv.com/developers/api#global-emotes
export async function fetchBetterTTVGlobalEmotes() {
  const response = await fetch(
    "https://api.betterttv.net/3/cached/emotes/global",
  );
  assert.response.ok(response, "Failed to load BetterTTV global emotes");
  const json = await response.json();
  return json as BetterTTVApi.GlobalEmote[];
}
