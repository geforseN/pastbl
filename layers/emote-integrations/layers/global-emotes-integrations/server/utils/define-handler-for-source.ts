import consola from "consola";
import type { GlobalEmoteIntegrationRecord } from "./get-all";

const consola_ = consola.withTag("global-emotes-integrations");

export function defineGlobalEmotesIntegrationEventHandler<S extends EmoteSource>(
  source: S,
) {
  const _integration = globalEmotesIntegrations.of(
    source,
  ) as GlobalEmoteIntegrationRecord[S];
  return async function (event: H3Event) {
    const { path } = event;
    consola_.debug("GET", { path });
    const integration = await _integration.get() as Awaited<ReturnType<GlobalEmoteIntegrationRecord[S]["get"]>>;
    consola_.debug("GET", {
      path,
      status: integration.status,
      source: integration.source,
    });
    return {
      integration,
    };
  };
}
