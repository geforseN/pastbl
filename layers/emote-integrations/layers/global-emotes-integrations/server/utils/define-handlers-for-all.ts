import { consola } from "consola";

export function defineGlobalEmotesIntegrationsEventHandlers(path: string) {
  const router = createRouter();
  for (const globalEmotesIntegration of globalEmotesIntegrations) {
    router.get(
      "/" + globalEmotesIntegration.source,
      defineEventHandler(async () => {
        const LOG_STR = "GET " + path + "/" + globalEmotesIntegration.source;
        consola.debug(LOG_STR);
        let integration: Awaited<
          ReturnType<typeof globalEmotesIntegration.get>
        >;
        try {
          integration = await globalEmotesIntegration.get();
          assert.ok(integration);
        } catch (reason) {
          const source = globalEmotesIntegration.source;
          return {
            source,
            status: "failed",
            reason: findErrorMessage(
              reason,
              `Failed to load ${source} Global Emote Integration`,
            ),
          };
        }
        consola.debug(LOG_STR + ":" + integration.status);
        return {
          integration,
        };
      }),
    );
  }
  return useBase(path, router.handler);
}
