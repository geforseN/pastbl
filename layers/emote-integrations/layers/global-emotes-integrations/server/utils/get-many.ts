export async function getGlobalEmotesIntegrations(sources: EmoteSource[]) {
  const values = await Promise.all(
    sources
      .toSorted()
      .map((source) => globalEmotesIntegrations.of(source).get()),
  );
  return flatGroupBySource(values);
}
