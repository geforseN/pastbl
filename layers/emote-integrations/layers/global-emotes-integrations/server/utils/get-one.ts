import type { EmoteSource } from "../../../emote-sources/utils/external";

export async function getGlobalEmotesIntegration(source: EmoteSource) {
  return await globalEmotesIntegrations.of(source).get();
}
