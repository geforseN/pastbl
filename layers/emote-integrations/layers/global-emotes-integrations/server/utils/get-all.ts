import { assert } from "~/utils/assert";
import type {
  BetterTTVGlobalEmotesIntegration,
  FrankerFaceZGlobalEmotesIntegration,
  SevenTVGlobalEmotesIntegration,
  TwitchGlobalEmotesIntegration,
} from "#build/types/nitro-imports";

export type GlobalEmoteIntegrationRecord = {
  BetterTTV: BetterTTVGlobalEmotesIntegration;
  FrankerFaceZ: FrankerFaceZGlobalEmotesIntegration;
  SevenTV: SevenTVGlobalEmotesIntegration;
  Twitch: TwitchGlobalEmotesIntegration;
};

const getters = [
  () => globalEmotesIntegrations.of("BetterTTV").get(),
  () => globalEmotesIntegrations.of("FrankerFaceZ").get(),
  () => globalEmotesIntegrations.of("SevenTV").get(),
  () => globalEmotesIntegrations.of("Twitch").get(),
] as const;

export async function getAllGlobalEmotesIntegrations() {
  const [BetterTTV, FrankerFaceZ, SevenTV, Twitch] = await Promise.all(
    getters.map((getter) => getter()),
  );

  assert.ok(BetterTTV && FrankerFaceZ && SevenTV && Twitch);

  return {
    BetterTTV,
    FrankerFaceZ,
    SevenTV,
    Twitch,
  };
}
