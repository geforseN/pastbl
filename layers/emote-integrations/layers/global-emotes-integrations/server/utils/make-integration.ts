import type { TEmoteIntegrations } from "$/emote-integrations";
import { EmotesIntegrationWithFailControl } from "$/emote-integrations/server/utils/fail-control";

export function defineGlobalIntegrationMaker<
  I extends TEmoteIntegrations.Global.Ready,
>(source: I["source"]) {
  return function (sets: I["sets"]) {
    return <I>{
      formedAt: Date.now(),
      sets,
      source,
      status: "ready" as const,
    };
  };
}

export class GlobalEmotesIntegrationWithFailControl extends EmotesIntegrationWithFailControl {
  constructor(readonly source: EmoteSource) {
    super();
  }

  override makeFailedIntegration<F extends FailedIntegration>(
    error: unknown,
  ): F {
    return <F>{
      status: "failed",
      source: this.source,
      code: "GLOBAL_EMOTES_INTEGRATION_FAILED",
      reason: findErrorMessage(
        error,
        `Failed to load ${this.source} Global Emotes Integration`,
      ),
    };
  }
}
