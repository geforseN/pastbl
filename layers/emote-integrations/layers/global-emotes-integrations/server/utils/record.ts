import { raise } from "../../../../../../app/utils/raise";
import type { EmoteSource } from "../../../emote-sources/utils/external";

export const globalEmotesIntegrations = {
  BetterTTV: new BetterTTVGlobalEmotesIntegration(
    new GlobalEmotesIntegrationWithFailControl("BetterTTV"),
  ),
  FrankerFaceZ: new FrankerFaceZGlobalEmotesIntegration(
    new GlobalEmotesIntegrationWithFailControl("FrankerFaceZ"),
  ),
  SevenTV: new SevenTVGlobalEmotesIntegration(
    new GlobalEmotesIntegrationWithFailControl("SevenTV"),
  ),
  Twitch: new TwitchGlobalEmotesIntegration(
    new GlobalEmotesIntegrationWithFailControl("Twitch"),
  ),
  of<T extends EmoteSource>(source: T) {
    return this[source] || raise();
  },
  *[Symbol.iterator]() {
    yield this.BetterTTV;
    yield this.FrankerFaceZ;
    yield this.SevenTV;
    yield this.Twitch;
  },
} as const;
