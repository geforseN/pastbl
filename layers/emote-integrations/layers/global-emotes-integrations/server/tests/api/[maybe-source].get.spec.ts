import { describe, it, expect } from "vitest";
import { integrationWithAnyFormedAtNumber, makeShortFrankerFaceZGlobalSet } from "./utils";
import { $apiFetch } from "~~/vitest/server-api/$apiFetch";

describe("GET /api/v1/global-emotes-integrations/[maybe-source]", () => {
  // FIXME
  // FIXME
  // FIXME
  it.todo.for([...allEmoteSources])("%s matches snapshot", async (source) => {
    const response = await $apiFetch("/global-emotes-integrations/" + source);
    if (source === "FrankerFaceZ") {
      response.integration.sets = response.integration.sets.map(makeShortFrankerFaceZGlobalSet);
    }
    expect(response).toMatchSnapshot({
      integration: integrationWithAnyFormedAtNumber,
    });
  });

  describe("404 on unknown non emote source param", async () => {
    const response = await $apiFetch.raw("/global-emotes-integrations/unknown", {
      ignoreResponseError: true,
    });

    it("will be not ok", () => {
      expect(response.ok).toBe(false);
    });

    it("will not be 400, but 404", () => {
      expect(response.status).not.toBe(400);
      expect(response.status).toBe(404);
    });

    it.todo("data will contain error message");
  });
});
