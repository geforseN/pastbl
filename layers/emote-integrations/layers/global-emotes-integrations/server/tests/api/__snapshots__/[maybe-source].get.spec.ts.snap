// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /api/v1/global-emotes-integrations/[maybe-source] > 404 on unknown non emote source param > _data > will match snapshot 1`] = `
{
  "data": {
    "path": "/api/v1/global-emotes-integrations/unknown",
  },
  "message": "Page not found: /api/v1/global-emotes-integrations/unknown",
  "stack": "",
  "statusCode": 404,
  "statusMessage": "Page not found: /api/v1/global-emotes-integrations/unknown",
  "url": "/api/v1/global-emotes-integrations/unknown",
}
`;
